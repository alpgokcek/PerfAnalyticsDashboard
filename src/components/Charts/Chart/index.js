import React from "react";
import PropTypes from "prop-types";
import { Line } from "react-chartjs-2";
import EmptyContent from "./EmptyContent";

import "./chart.scss";

const Chart = (props) => {
  const { title, keyField, metrics } = props;

  const categorizedMetrics = metrics.reduce(function (r, a) {
    r[a.url] = r[a.url] || [];
    r[a.url].push(a);
    return r;
  }, Object.create(null));

  const datasets = Object.keys(categorizedMetrics).map((k) => {
    return {
      label: k,
      data: categorizedMetrics[k].map((item) => ({
        x: item.timestamp,
        y: item[keyField],
      })),
    };
  });
  const data = {
    labels: metrics.map((item) => item.timestamp).sort(),
    datasets,
  };

  return (
    <div className="chart">
      <div className="chart-title" data-test="chart-title">
        {title}
      </div>
      <div data-test={`chart-${keyField}-wrapper`}>
        {metrics.length > 0 ? (
          <Line
            className="chart-canvas"
            data={data}
            data-test={`chart-${keyField}`}
          />
        ) : (
          <EmptyContent data-test="chart-empty" />
        )}
      </div>
    </div>
  );
};

Chart.propTypes = {
  title: PropTypes.string.isRequired,
  keyField: PropTypes.string.isRequired,
  metrics: PropTypes.array.isRequired,
};

export default Chart;
