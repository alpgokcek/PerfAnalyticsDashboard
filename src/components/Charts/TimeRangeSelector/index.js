import React from "react";
import PropTypes from "prop-types";
import { DateTimePickerComponent } from "@syncfusion/ej2-react-calendars";

import "./time-range-selector.scss";

const TimeRangeSelector = (props) => {
  const { onChange, filters } = props;

  const handleDateChange = (e, isStartDate = false) => {
    if (isStartDate) onChange({ ...filters, start: e.target.value });
    else onChange({ ...filters, end: e.target.value });
  };

  return (
    <div className="time-range-selector">
      <div className="time-range-selector__container">
        <div>Start Date</div>
        <DateTimePickerComponent
          data-test="datetime-start"
          placeholder="Choose a start date and time"
          value={filters.start}
          onChange={(e) => handleDateChange(e, true)}
          format="dd-MMM-yy HH:mm"
          step={30}
        ></DateTimePickerComponent>
      </div>

      <div className="time-range-selector__container">
        <div>End Date</div>
        <DateTimePickerComponent
          data-test="datetime-end"
          placeholder="Choose an end date and time"
          value={filters.end}
          onChange={(e) => handleDateChange(e)}
          format="dd-MMM-yy HH:mm"
          step={30}
        ></DateTimePickerComponent>
      </div>
    </div>
  );
};

TimeRangeSelector.propTypes = {
  onChange: PropTypes.func,
  filters: PropTypes.object,
};

export default TimeRangeSelector;
