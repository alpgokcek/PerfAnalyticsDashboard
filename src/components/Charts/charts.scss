@import "../../styles/variables.scss";
@import "../../styles/placeholders.scss";

.charts {
  &__container {
    display: grid;
    grid-template-columns: auto auto;
    column-gap: 20px;
    row-gap: 20px;
    margin-top: 40px;
    @include tablet {
      grid-template-columns: auto;
    }
    @include phone {
      grid-template-columns: auto;
    }
  }
  &__time-range {
    &-reset {
      display: inline-block;
      border: 1px solid $colorCharcoal;
      padding: 10px;
      font-size: 14px;
      margin-top: 20px;
    }
    &-cta-wrapper {
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }
}
