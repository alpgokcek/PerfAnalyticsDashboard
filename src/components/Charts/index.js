import React from "react";
import PropTypes from "prop-types";
import Chart from "./Chart";
import { connect } from "react-redux";
import { getMetrics } from "../../store/actions";
import TimeRangeSelector from "./TimeRangeSelector";

import "./charts.scss";

const Charts = (props) => {
  const { metrics } = props;
  const [queryFilters, setQueryFilters] = React.useState({
    start: new Date(),
    end: new Date(),
  });
  React.useEffect(() => {
    props.getMetrics(
      queryFilters.start === queryFilters.end ? {} : queryFilters
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [queryFilters]);

  const resetFilters = () => {
    setQueryFilters({
      start: new Date(),
      end: new Date(),
    });
  };

  const ttfbChart = <Chart title="TTFB" keyField="ttfb" metrics={metrics} />;
  const fcpChart = <Chart title="FCP" keyField="fcp" metrics={metrics} />;
  const domLoadChart = (
    <Chart title="DOM Load" keyField="domLoad" metrics={metrics} />
  );
  const windowLoadEventsChart = (
    <Chart
      title="Window Load Events"
      keyField="windowLoadEvents"
      metrics={metrics}
    />
  );

  return (
    <div>
      <div className="charts__time-range">
        <TimeRangeSelector
          data-test="charts-datetime"
          onChange={setQueryFilters}
          filters={queryFilters}
        />
        <div className="charts__time-range-cta-wrapper">
          <div
            className="charts__time-range-reset"
            data-test="charts-range-reset"
            onClick={resetFilters}
          >
            Reset Range
          </div>
        </div>
      </div>
      <div className="charts__container" data-test="charts-container">
        {ttfbChart}
        {fcpChart}
        {domLoadChart}
        {windowLoadEventsChart}
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    metrics: state.metricsReducer.metrics,
  };
};
const mapDispatchToProps = (dispatch) => {
  return {
    getMetrics: (filters) => {
      dispatch(getMetrics(filters));
    },
  };
};

Charts.propTypes = {
  getMetrics: PropTypes.func,
  metrics: PropTypes.array,
};

export default connect(mapStateToProps, mapDispatchToProps)(Charts);
