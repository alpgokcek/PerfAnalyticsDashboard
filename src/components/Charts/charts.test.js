import React from "react";
import Charts from "./index";
import * as redux from "react-redux";

import Enzyme, { mount } from "enzyme";
import toJson from "enzyme-to-json";
import EnzymeAdapter from "enzyme-adapter-react-16";
import { findByTestAttribute, mockStore } from "../../test";
import { MOCK_METRIC } from "../../test/consts";
import { Provider } from "react-redux";
import moment from "moment";

Enzyme.configure({ adapter: new EnzymeAdapter() });

const defaultProps = {
  getMetrics: jest.fn,
  metrics: [MOCK_METRIC, MOCK_METRIC, MOCK_METRIC],
};

const setup = (store, props) => {
  return mount(
    <Provider store={store}>
      <Charts {...props} />
    </Provider>
  );
};

jest.mock("react-chartjs-2", () => ({
  Line: () => <div>chart</div>,
}));

describe("snapshot and unit tests for charts", () => {
  let wrapper, useDispatchSpy;
  beforeAll(() => {
    const mGetRandomValues = jest.fn().mockReturnValueOnce(new Uint32Array(10));
    Object.defineProperty(window, "crypto", {
      value: { getRandomValues: mGetRandomValues },
    });
    jest
      .useFakeTimers("modern")
      .setSystemTime(new Date("2021-05-23").getTime());
  });

  beforeEach(() => {
    const store = mockStore({});
    wrapper = setup(store, defaultProps);
    useDispatchSpy = jest.spyOn(store, "dispatch");
    useDispatchSpy.mockReturnValue(jest.fn());
  });
  afterEach(() => {
    useDispatchSpy.mockClear();
  });

  it("should render correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render TimeRangeSelector", () => {
    const inputComponent = findByTestAttribute(wrapper, "charts-datetime");
    expect(inputComponent.exists()).toBeTruthy();
  });

  it("should reset datetime range", () => {
    const date = new Date("2021-05-25").getTime();
    const expectedDateString = moment(date).format("DD-MMM-YY HH:mm");
    jest.useFakeTimers("modern").setSystemTime(date);
    findByTestAttribute(wrapper, "charts-range-reset")
      .first()
      .simulate("click");
    const startDate = findByTestAttribute(wrapper, "datetime-start");
    expect(startDate.find("input").instance().value).toMatch(
      expectedDateString
    );
    const endDate = findByTestAttribute(wrapper, "datetime-end");
    expect(endDate.find("input").instance().value).toMatch(expectedDateString);
  });

  it("should render charts correctly", () => {
    wrapper.setProps({ ...defaultProps });
    let inputComponent = findByTestAttribute(wrapper, "charts-container");
    expect(inputComponent.children()).toHaveLength(4);
  });

  it("should not render chart", () => {
    wrapper.setProps({ ...defaultProps, metrics: [] });
    let inputComponent = findByTestAttribute(wrapper, "chart-ttfb-wrapper");
    expect(inputComponent.children()).toHaveLength(1);
    inputComponent = findByTestAttribute(wrapper, "chart-ttfb");
    expect(inputComponent.exists()).not.toBeTruthy();
    inputComponent = findByTestAttribute(wrapper, "chart-empty");
    expect(inputComponent.exists()).toBeTruthy();
  });
});
