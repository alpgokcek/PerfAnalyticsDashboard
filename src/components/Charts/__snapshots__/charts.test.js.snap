// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot and unit tests for charts should render correctly 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [MockFunction],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Charts)
    getMetrics={[Function]}
    metrics={
      Array [
        Object {
          "domLoad": "200",
          "fcp": "500",
          "timestamp": "2021-05-22T22:33:10.292Z",
          "ttfb": "700",
          "url": "https://google.com",
          "windowLoadEvents": "300",
        },
        Object {
          "domLoad": "200",
          "fcp": "500",
          "timestamp": "2021-05-22T22:33:10.292Z",
          "ttfb": "700",
          "url": "https://google.com",
          "windowLoadEvents": "300",
        },
        Object {
          "domLoad": "200",
          "fcp": "500",
          "timestamp": "2021-05-22T22:33:10.292Z",
          "ttfb": "700",
          "url": "https://google.com",
          "windowLoadEvents": "300",
        },
      ]
    }
  >
    <Charts
      getMetrics={[Function]}
      metrics={Array []}
    >
      <div>
        <div
          className="charts__time-range"
        >
          <TimeRangeSelector
            data-test="charts-datetime"
            filters={
              Object {
                "end": 2021-05-23T00:00:00.000Z,
                "start": 2021-05-23T00:00:00.000Z,
              }
            }
            onChange={[Function]}
          >
            <div
              className="time-range-selector"
            >
              <div
                className="time-range-selector__container"
              >
                <div>
                  Start Date
                </div>
                <n
                  data-test="datetime-start"
                  format="dd-MMM-yy HH:mm"
                  onChange={[Function]}
                  placeholder="Choose a start date and time"
                  step={30}
                  value={2021-05-23T00:00:00.000Z}
                >
                  <input
                    data-test="datetime-start"
                    key="1"
                  />
                </n>
              </div>
              <div
                className="time-range-selector__container"
              >
                <div>
                  End Date
                </div>
                <n
                  data-test="datetime-end"
                  format="dd-MMM-yy HH:mm"
                  onChange={[Function]}
                  placeholder="Choose an end date and time"
                  step={30}
                  value={2021-05-23T00:00:00.000Z}
                >
                  <input
                    data-test="datetime-end"
                    key="2"
                  />
                </n>
              </div>
            </div>
          </TimeRangeSelector>
          <div
            className="charts__time-range-cta-wrapper"
          >
            <div
              className="charts__time-range-reset"
              data-test="charts-range-reset"
              onClick={[Function]}
            >
              Reset Range
            </div>
          </div>
        </div>
        <div
          className="charts__container"
          data-test="charts-container"
        >
          <Chart
            keyField="ttfb"
            metrics={Array []}
            title="TTFB"
          >
            <div
              className="chart"
            >
              <div
                className="chart-title"
                data-test="chart-title"
              >
                TTFB
              </div>
              <div
                data-test="chart-ttfb-wrapper"
              >
                <EmptyContent
                  data-test="chart-empty"
                >
                  <div
                    className="empty-content"
                  >
                    <div
                      className="empty-content__wrapper"
                    >
                      <img
                        alt="no-data"
                        className="empty-content__wrapper-icon"
                        src="assets/img/no-data.png"
                      />
                      <div>
                        Not content found.
                      </div>
                    </div>
                  </div>
                </EmptyContent>
              </div>
            </div>
          </Chart>
          <Chart
            keyField="fcp"
            metrics={Array []}
            title="FCP"
          >
            <div
              className="chart"
            >
              <div
                className="chart-title"
                data-test="chart-title"
              >
                FCP
              </div>
              <div
                data-test="chart-fcp-wrapper"
              >
                <EmptyContent
                  data-test="chart-empty"
                >
                  <div
                    className="empty-content"
                  >
                    <div
                      className="empty-content__wrapper"
                    >
                      <img
                        alt="no-data"
                        className="empty-content__wrapper-icon"
                        src="assets/img/no-data.png"
                      />
                      <div>
                        Not content found.
                      </div>
                    </div>
                  </div>
                </EmptyContent>
              </div>
            </div>
          </Chart>
          <Chart
            keyField="domLoad"
            metrics={Array []}
            title="DOM Load"
          >
            <div
              className="chart"
            >
              <div
                className="chart-title"
                data-test="chart-title"
              >
                DOM Load
              </div>
              <div
                data-test="chart-domLoad-wrapper"
              >
                <EmptyContent
                  data-test="chart-empty"
                >
                  <div
                    className="empty-content"
                  >
                    <div
                      className="empty-content__wrapper"
                    >
                      <img
                        alt="no-data"
                        className="empty-content__wrapper-icon"
                        src="assets/img/no-data.png"
                      />
                      <div>
                        Not content found.
                      </div>
                    </div>
                  </div>
                </EmptyContent>
              </div>
            </div>
          </Chart>
          <Chart
            keyField="windowLoadEvents"
            metrics={Array []}
            title="Window Load Events"
          >
            <div
              className="chart"
            >
              <div
                className="chart-title"
                data-test="chart-title"
              >
                Window Load Events
              </div>
              <div
                data-test="chart-windowLoadEvents-wrapper"
              >
                <EmptyContent
                  data-test="chart-empty"
                >
                  <div
                    className="empty-content"
                  >
                    <div
                      className="empty-content__wrapper"
                    >
                      <img
                        alt="no-data"
                        className="empty-content__wrapper-icon"
                        src="assets/img/no-data.png"
                      />
                      <div>
                        Not content found.
                      </div>
                    </div>
                  </div>
                </EmptyContent>
              </div>
            </div>
          </Chart>
        </div>
      </div>
    </Charts>
  </Connect(Charts)>
</Provider>
`;
